src/state/internal/createQueryStore.ts

```diff
@@ -0,0 +1,775 @@
+    params?: {
+      [K in keyof TParams]: ParamResolvable<TParams[K]>;
+    };
+  },
+  customStateCreator?: StateCreator<U, [], [['zustand/subscribeWithSelector', never]]>,
+  persistConfig?: RainbowPersistConfig<StoreState<TData, TParams> & U>
+): QueryStore<TData, TParams, StoreState<TData, TParams> & U> {
+  type S = StoreState<TData, TParams> & U;
+
+  const {
+    fetcher,
+    onFetched,
+    transform,
+    cacheTime = time.days(7),
+    disableCache = false,
+    enabled = true,
+    maxRetries = 3,
+    onError,
+    params,
+    retryDelay = time.seconds(5),
+    setData,
+    staleTime = time.minutes(2),
+    suppressStaleTimeWarning = false,
+  } = config;
+
+  if (IS_DEV && !suppressStaleTimeWarning && staleTime < MIN_STALE_TIME) {
+    console.warn(
+      `[RainbowQueryStore${persistConfig?.storageKey ? `: ${persistConfig.storageKey}` : ''}] ❌ Stale times under ${
+        MIN_STALE_TIME / 1000
+      } seconds are not recommended.`
+    );
+  }
+
+  let directValues: Partial<TParams> = {};
+  let paramAttachVals: Partial<Record<keyof TParams, AttachValue<unknown>>> = {};
+
+  if (params) {
+    const result = resolveParams<TParams>(params);
+    paramAttachVals = result.paramAttachVals;
+    directValues = result.directValues;
+  }
+
+  let activeFetchPromise: Promise<void> | null = null;
+  let activeRefetchTimeout: NodeJS.Timeout | null = null;
+  let lastFetchKey: string | null = null;
+
+  const initialData = {
+    enabled,
+    error: null,
+    lastFetchedAt: null,
+    queryCache: {},
+    status: QueryStatuses.Idle,
+    subscriptionCount: 0,
+  };
+
+  const getQueryKey = (params: TParams): string => JSON.stringify(Object.values(params));
+
+  const getCurrentResolvedParams = () => {
+    const currentParams = { ...directValues };
+    for (const k in paramAttachVals) {
+      const attachVal = paramAttachVals[k as keyof TParams];
+      if (!attachVal) continue;
+      currentParams[k as keyof TParams] = attachVal.value as TParams[keyof TParams];
+    }
+    return currentParams as TParams;
+  };
+
+  const createState: StateCreator<S, [], [['zustand/subscribeWithSelector', never]]> = (set, get, api) => {
+    const pruneCache = (state: S): S => {
+      const now = Date.now();
+      const newCache: Record<string, CacheEntry<TData>> = {};
+      Object.entries(state.queryCache).forEach(([key, entry]) => {
+        if (entry && now - entry.lastFetchedAt <= cacheTime) {
+          newCache[key] = entry;
+        }
+      });
+      return { ...state, queryCache: newCache };
+    };
+
+    const scheduleNextFetch = (params: TParams, options: FetchOptions | undefined) => {
+      const effectiveStaleTime = options?.staleTime ?? staleTime;
+      if (effectiveStaleTime <= 0 || effectiveStaleTime === Infinity) return;
+      if (activeRefetchTimeout) {
+        clearTimeout(activeRefetchTimeout);
+        activeRefetchTimeout = null;
+      }
```

Comment:
after taking a glance at your example, it seems like a separate store might set up for a mutually exclusive set of queries?

Reply:
Right now it’s a query-per-store model — each query store operates independently, though a store's query params can be defined as dynamic subscriptions to any store’s state using the `$` API.

It would be pretty straightforward to support multiple queries, since the caching system is already set up to handle multiple keys — just wanted to keep it simple initially and get all the core functionality working. I do think the single-query model may still be favorable since it encourages keeping stores small and modular, which helps limit serialization overhead.